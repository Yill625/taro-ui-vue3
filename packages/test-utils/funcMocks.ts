import * as utils from "@taro-ui-vue3/utils/common"
import { pxTransformMockFn, getEnvMockFn } from './@tarojs/taro'

let uuid: jest.SpyInstance
let getEnvs: jest.SpyInstance
let pxTransform: jest.SpyInstance

function initMocks() {
  beforeEach(() => {
    jest.mock('@taro-ui-vue3/utils/common', () => {
      return utils
    })

    pxTransform = jest
      .spyOn(utils, 'pxTransform')
      .mockImplementation(pxTransformMockFn)

    uuid = jest
      .spyOn(utils, 'uuid')
      .mockReturnValue('__tests__2021')

    getEnvs = jest
      .spyOn(utils, 'getEnvs')
      .mockImplementation(() => {
        const env = getEnvMockFn()
        return {
          isWEAPP: env === 'WEAPP',
          isALIPAY: env === 'ALIPAY',
          isWEB: env === 'WEB'
        }
      })
  })

  afterEach(() => {
    uuid.mockRestore()
    getEnvs.mockRestore()
    pxTransform.mockRestore()
  })
}

export default initMocks