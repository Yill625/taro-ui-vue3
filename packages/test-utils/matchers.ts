import * as utils from "@taro-ui-vue3/utils/common"

jest.mock('@taro-ui-vue3/utils/common')

let warn: jest.SpyInstance
let uuid: jest.SpyInstance

const asserted: Set<string> = new Set()
const NOOP = () => { }

// Ensure console.warn is called with expected message
const toHaveBeenWarned: jest.CustomMatcher = (received: string) => {
  asserted.add(received)
  const passed = warn.mock.calls.some(args => args[0].indexOf(received) > -1)
  if (passed) {
    return {
      pass: true,
      message: () => `expected "${received}" not to have been warned.`
    }
  } else {
    const msgs = warn.mock.calls.map(args => args[0]).join('\n - ')
    return {
      pass: false,
      message: () =>
        `expected "${received}" to have been warned` +
        (msgs.length
          ? `.\n\nActual messages:\n\n - ${msgs}`
          : ` but no warning was recorded.`)
    }
  }
}

// Ensure previous console.warn is called with expected message
const toHaveBeenWarnedLast: jest.CustomMatcher = (received: string) => {
  asserted.add(received)
  const passed =
    warn.mock.calls[warn.mock.calls.length - 1][0].indexOf(received) > -1
  if (passed) {
    return {
      pass: true,
      message: () => `expected "${received}" not to have been warned last.`
    }
  } else {
    const msgs = warn.mock.calls.map(args => args[0]).join('\n - ')
    return {
      pass: false,
      message: () =>
        `expected "${received}" to have been warned last.\n\nActual messages:\n\n - ${msgs}`
    }
  }
}

// Ensure console.warn is called n times with expected message
const toHaveBeenWarnedTimes: jest.CustomMatcher = (received: string, n: number) => {
  asserted.add(received)
  let found = 0
  warn.mock.calls.forEach(args => {
    if (args[0].indexOf(received) > -1) {
      found++
    }
  })

  if (found === n) {
    return {
      pass: true,
      message: () => `expected "${received}" to have been warned ${n} times.`
    }
  } else {
    return {
      pass: false,
      message: () =>
        `expected "${received}" to have been warned ${n} times but got ${found}.`
    }
  }
}

function initWarnMatchers() {
  beforeEach(() => {
    asserted.clear()
    warn = jest.spyOn(console, 'warn').mockImplementation(NOOP)
    uuid = jest.spyOn(utils, 'uuid').mockReturnValue('__tests__2021')

    // NOT working in concurrent mode
    // should avoid use in test.concurrent()
    expect.extend({
      toHaveBeenWarned,
      toHaveBeenWarnedLast,
      toHaveBeenWarnedTimes
    })
  })

  afterEach(() => {
    const assertedArray = Array.from(asserted)
    const nonAssertedWarnings = warn.mock.calls
      .map(args => args[0])
      .filter(received => {
        return !assertedArray.some(assertedMsg => {
          return received.indexOf(assertedMsg) > -1
        })
      })

    uuid.mockRestore()
    warn.mockRestore()

    if (nonAssertedWarnings.length) {
      throw new Error(
        `test case threw unexpected warnings:\n - ${nonAssertedWarnings.join(
          '\n - '
        )}`
      )
    }
  })
}

export default initWarnMatchers