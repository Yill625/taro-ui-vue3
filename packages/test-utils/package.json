{
  "name": "@taro-ui-vue3/test-utils",
  "version": "0.0.0",
  "main": "dist/index.js",
  "license": "MIT",
  "dependencies": {
    "vue": "^3.0.10",
    "@vue/test-utils": "^2.0.0-rc.0"
  }
}