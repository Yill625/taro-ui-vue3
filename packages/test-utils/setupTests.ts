import toHaveBeenWarnedInit from './been-warned'

toHaveBeenWarnedInit()