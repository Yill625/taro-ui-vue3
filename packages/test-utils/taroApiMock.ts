import * as utils from "@taro-ui-vue3/utils/common"
import Taro from '@tarojs/taro'
import { createAnimation } from '@tarojs/taro-h5'

import {
  getEnvMockFn,
  pxTransformMockFn,
  chooseImageMockFn,
  querySelectorMockFn
} from './@tarojs/taro'

Taro.pxTransform = pxTransformMockFn
Taro.chooseImage = chooseImageMockFn
Taro.createSelectorQuery = querySelectorMockFn()
Taro.createAnimation = createAnimation

let uuid: jest.SpyInstance
let getEnvs: jest.SpyInstance
let getenv: jest.SpyInstance
let pxTransform: jest.SpyInstance
let chooseImage: jest.SpyInstance

function initTaroMockApis() {
  beforeEach(() => {
    jest.mock('@tarojs/taro', () => {
      return Taro
    })

    jest.mock('@taro-ui-vue3/utils/common', () => {
      return utils
    })

    getenv = jest
      .spyOn(Taro, 'getEnv')
      .mockImplementation(getEnvMockFn)

    chooseImage = jest
      .spyOn(Taro, 'chooseImage')
      .mockImplementation(chooseImageMockFn)

    pxTransform = jest
      .spyOn(utils, 'pxTransform')
      .mockImplementation(pxTransformMockFn)

    uuid = jest
      .spyOn(utils, 'uuid')
      .mockReturnValue('__tests__2021')

    getEnvs = jest
      .spyOn(utils, 'getEnvs')
      .mockImplementation(() => {
        const env = getEnvMockFn()
        return {
          isWEAPP: env === 'WEAPP',
          isALIPAY: env === 'ALIPAY',
          isWEB: env === 'WEB'
        }
      })
  })

  afterEach(() => {
    uuid.mockRestore()
    getEnvs.mockRestore()
    getenv.mockRestore()
    pxTransform.mockRestore()
    chooseImage.mockRestore()
  })
}

export default initTaroMockApis
