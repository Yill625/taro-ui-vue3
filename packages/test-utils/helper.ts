import { mount, shallowMount, DOMWrapper, VueWrapper } from '@vue/test-utils'
import { VNode, Slot } from '@vue/runtime-core'

export const sleep = async (timeout: number): Promise<null> =>
  new Promise((resolve) => setTimeout(resolve, timeout))

export interface Slots {
  default: Array<string | VNode> | Slot
  [x: string]: Array<string | VNode> | Slot
}

export function mountFactory(
  comp: any,
  components?: Record<string, object>,
  props?: Record<string, any>,
  slots?: Slots
) {
  return mount<any>(comp, {
    global: {
      components
    },
    props,
    slots,
  })
}

export function genMountFn(
  comp: any,
  components?: Record<string, object>,
  useShallowMount?: boolean
) {
  return (props = {}, slots?: Slots) => {
    return useShallowMount
      ? shallowMountFactory(comp, components, props, slots)
      : mountFactory(comp, components, props, slots)
  }
}

export function shallowMountFactory(
  comp: any,
  components?: Record<string, object>,
  props?: Record<string, any>,
  slots?: Slots
) {
  return shallowMount<any>(comp, {
    global: {
      components
    },
    props,
    slots,
  })
}

export interface Dimension {
  top?: number
  bottom?: number
  left?: number
  right?: number
  width?: number
  height?: number
}

export function genDelayedSelectorSpy(
  utils: any,
  dimensions: Dimension
): jest.SpyInstance {
  return jest
    .spyOn(utils, 'delayQuerySelector')
    .mockImplementation((_, sel: string, delay?: number) => {
      return new Promise((resolve) => {
        resolve([dimensions])
      })
    })
}

export async function triggerTouchEvents(
  el: DOMWrapper<Element>,
  touchstartDetails: any,
  touchendDetails: any
) {
  const startDetails = {
    touches: [touchstartDetails],
    changedTouches: [touchstartDetails]
  }

  const endDetails = {
    touches: [touchendDetails],
    changedTouches: [touchendDetails]
  }

  await el.trigger('touchstart', startDetails)
  await el.trigger('touchmove', endDetails)
  await el.trigger('touchend', endDetails)
}

export function testPropClassAndStyle(
  mountFn: (props?: {}, slots?: Slots | undefined) => VueWrapper<any>
) {

  it.concurrent('should render prop class', async () => {
    const wrapper = mountFn({ class: 'test' })
    expect(
      wrapper.classes()
    ).toContain('test')
  })

  it.concurrent('should render prop style', async () => {
    const wrapper = mountFn({ style: { color: 'red' } })
    expect(
      wrapper.attributes('style')
    ).toContain('color: red;')
  })
}

export function hexToRGBA(hex: string, opacity = 1) {
  hex = hex.replace('#', '')
  const rgb = hex
    .match(new RegExp('(.{' + hex.length / 3 + '})', 'g'))!
    .map((l) => parseInt(hex.length % 2 ? l + l : l, 16))

  const rgba = rgb.concat(isFinite(opacity) ? opacity : 1)

  return opacity === 1
    ? `rgb(${rgb.join(', ')})`
    : `rgba(${rgba.join(', ')})`
}