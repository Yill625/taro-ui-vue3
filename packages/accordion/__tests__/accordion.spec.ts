import { mountFactory, Slots } from '@taro-ui-vue3/test-utils/helper'
import AtAccordion from '../index'

const factory = (
  values = {},
  slots: Slots = { default: ['按钮'] }
) => {
  return mountFactory(AtAccordion, {}, values, slots)
}

describe('AtAccordion', () => {
  it('should render default AtAccordion', () => {
    const wrapper = factory()
    expect(wrapper.element).toMatchSnapshot()
  })

  it('should render prop -- title', () => {
    const wrapper = factory({
      title: 'title',
    })
    expect(wrapper.element).toMatchSnapshot()
    expect(
      wrapper.find('.at-accordion__info__title').text()
    ).toEqual('title')
  })

  it('should render prop -- open=true', () => {
    const wrapper = factory({
      open: true,
    })

    expect(wrapper.element).toMatchSnapshot()
    expect(wrapper.find(".at-accordion__arrow--folded").exists()).toBeTruthy()
    expect(wrapper.find(".at-accordion__content--inactive").exists()).toBeFalsy()
  })

  it('should render prop -- open=false', () => {
    const wrapper = factory()
    expect(wrapper.element).toMatchSnapshot()
    expect(wrapper.find(".at-accordion__arrow--folded").exists()).toBeFalsy()
    expect(wrapper.find(".at-accordion__content--inactive").exists()).toBeTruthy()
  })

  it('should render prop -- icon', () => {
    const wrapper = factory({
      icon: { value: 'chevron-down' },
    })
    expect(wrapper.element).toMatchSnapshot()
    expect(wrapper.find(".at-icon-chevron-down").exists()).toBeTruthy()
  })

  it('should render prop -- icon with color', async () => {
    const wrapper = factory({
      icon: { value: 'chevron-down', color: 'red' },
    })
    expect(wrapper.element).toMatchSnapshot()
    expect(wrapper.find(".at-icon-chevron-down").exists()).toBeTruthy()
    expect(
      wrapper.find(".at-icon-chevron-down").attributes("style")
    ).toBe("color: red;")

    await wrapper.setProps({ icon: { value: 'chevron-down' } })
    wrapper.vm.$nextTick()
    expect(wrapper.find(".at-icon-chevron-down").exists()).toBeTruthy()
    expect(
      wrapper.find(".at-icon-chevron-down").attributes("style")
    ).toBe("")
  })

  it('should render prop -- icon with size', async () => {
    const wrapper = factory({
      icon: { value: 'chevron-down', size: 10 },
    })
    expect(wrapper.element).toMatchSnapshot()
    expect(wrapper.find(".at-icon-chevron-down").exists()).toBeTruthy()
    expect(
      wrapper.find(".at-icon-chevron-down").attributes().style
    ).toBe("font-size: 10px;")

    await wrapper.setProps({ icon: { value: 'chevron-down' } })
    wrapper.vm.$nextTick()
    expect(wrapper.find(".at-icon-chevron-down").exists()).toBeTruthy()
    expect(
      wrapper.find(".at-icon-chevron-down").attributes("style")
    ).toBe("")
  })

  it('should render prop -- icon with prefixClass', async () => {
    const wrapper = factory({
      icon: { prefixClass: 'prefixClass', value: 'star' },
    })
    expect(wrapper.element).toMatchSnapshot()

    const iconElement = wrapper.find(".at-accordion__icon")
    expect(iconElement.classes()).toContain('prefixClass')
    expect(iconElement.classes()).toContain('prefixClass-star')

    await wrapper.setProps({ icon: { value: 'star' } })
    wrapper.vm.$nextTick()
    const iconElement2 = wrapper.find(".at-accordion__icon")
    expect(iconElement2.classes()).not.toContain('prefixClass')
    expect(iconElement2.classes()).not.toContain('prefixClass-star')
  })

  it('should render prop -- hasBorder', async () => {
    const wrapper = factory({
      title: 'title',
    })

    expect(wrapper.find(".at-accordion__header--noborder").exists()).toBeFalsy()

    await wrapper.setProps({ hasBorder: false })
    wrapper.vm.$nextTick()
    expect(wrapper.find(".at-accordion__header--noborder").exists()).toBeTruthy()
  })

  it('should render prop -- note', () => {
    const wrapper = factory({
      note: 'note',
    })
    expect(wrapper.element).toMatchSnapshot()
    expect(
      wrapper.find('.at-accordion__info__note').text()
    ).toEqual('note')
  })
})

describe('AtAccordion Behavior', () => {
  it('should trigger onClick event to toggle accordion', async () => {
    const onClick = jest.fn()
    const wrapper = factory({
      onClick: onClick,
    })

    expect(wrapper.find('.at-accordion__content--inactive').exists()).toBeTruthy()

    await wrapper.find('.at-accordion__header').trigger('tap')
    wrapper.vm.$nextTick(() => {
      expect(wrapper.find('.at-accordion__content--inactive').exists()).toBeFalsy()
    })

    await wrapper.find('.at-accordion__header').trigger('tap')
    wrapper.vm.$nextTick(() => {
      expect(wrapper.find('.at-accordion__content--inactive').exists()).toBeTruthy()
    })

    expect(onClick.mock.calls.length).toBe(2)
  })

  it('should not trigger onClick event when isAnimation is false', async () => {
    const onClick = jest.fn()
    const wrapper = factory({
      onClick: onClick,
    })

    await wrapper.setProps({ open: true, isAnimation: false })
    wrapper.vm.$nextTick(async () => {
      await wrapper.find('.at-accordion__header').trigger('tap')
      expect(onClick).not.toBeCalled()
    })
  })

  jest.useFakeTimers()

  it('should trigger setTimeout', async () => {
    jest.spyOn(global, 'setTimeout')
    const onClick = jest.fn()
    const wrapper = factory({
      onClick: onClick,
    })

    await wrapper.setProps({ open: true, isAnimation: true })
    jest.runAllTimers()
    expect(setTimeout).toHaveBeenNthCalledWith(1, expect.any(Function), 30)
  })

})
