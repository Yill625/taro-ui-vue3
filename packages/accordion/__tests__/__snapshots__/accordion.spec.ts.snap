// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AtAccordion should render default AtAccordion 1`] = `
<view
  class="at-accordion"
>
  <view
    class="at-accordion__header"
  >
    
    <view
      class="at-accordion__info"
    >
      <view
        class="at-accordion__info__title"
      />
      <view
        class="at-accordion__info__note"
      />
    </view>
    <view
      class="at-accordion__arrow"
    >
      <text
        class="at-icon at-icon-chevron-down"
      />
    </view>
  </view>
  <view
    class="at-accordion__content at-accordion__content--inactive"
  >
    <view
      class="at-accordion__body"
      id="content"
    >
      按钮
    </view>
  </view>
</view>
`;

exports[`AtAccordion should render prop -- icon 1`] = `
<view
  class="at-accordion"
>
  <view
    class="at-accordion__header"
  >
    <text
      class="at-icon at-icon-chevron-down at-accordion__icon"
    />
    <view
      class="at-accordion__info"
    >
      <view
        class="at-accordion__info__title"
      />
      <view
        class="at-accordion__info__note"
      />
    </view>
    <view
      class="at-accordion__arrow"
    >
      <text
        class="at-icon at-icon-chevron-down"
      />
    </view>
  </view>
  <view
    class="at-accordion__content at-accordion__content--inactive"
  >
    <view
      class="at-accordion__body"
      id="content"
    >
      按钮
    </view>
  </view>
</view>
`;

exports[`AtAccordion should render prop -- icon with color 1`] = `
<view
  class="at-accordion"
>
  <view
    class="at-accordion__header"
  >
    <text
      class="at-icon at-icon-chevron-down at-accordion__icon"
      style="color: red;"
    />
    <view
      class="at-accordion__info"
    >
      <view
        class="at-accordion__info__title"
      />
      <view
        class="at-accordion__info__note"
      />
    </view>
    <view
      class="at-accordion__arrow"
    >
      <text
        class="at-icon at-icon-chevron-down"
      />
    </view>
  </view>
  <view
    class="at-accordion__content at-accordion__content--inactive"
  >
    <view
      class="at-accordion__body"
      id="content"
    >
      按钮
    </view>
  </view>
</view>
`;

exports[`AtAccordion should render prop -- icon with prefixClass 1`] = `
<view
  class="at-accordion"
>
  <view
    class="at-accordion__header"
  >
    <text
      class="prefixClass prefixClass-star at-accordion__icon"
    />
    <view
      class="at-accordion__info"
    >
      <view
        class="at-accordion__info__title"
      />
      <view
        class="at-accordion__info__note"
      />
    </view>
    <view
      class="at-accordion__arrow"
    >
      <text
        class="at-icon at-icon-chevron-down"
      />
    </view>
  </view>
  <view
    class="at-accordion__content at-accordion__content--inactive"
  >
    <view
      class="at-accordion__body"
      id="content"
    >
      按钮
    </view>
  </view>
</view>
`;

exports[`AtAccordion should render prop -- icon with size 1`] = `
<view
  class="at-accordion"
>
  <view
    class="at-accordion__header"
  >
    <text
      class="at-icon at-icon-chevron-down at-accordion__icon"
      style="font-size: 10px;"
    />
    <view
      class="at-accordion__info"
    >
      <view
        class="at-accordion__info__title"
      />
      <view
        class="at-accordion__info__note"
      />
    </view>
    <view
      class="at-accordion__arrow"
    >
      <text
        class="at-icon at-icon-chevron-down"
      />
    </view>
  </view>
  <view
    class="at-accordion__content at-accordion__content--inactive"
  >
    <view
      class="at-accordion__body"
      id="content"
    >
      按钮
    </view>
  </view>
</view>
`;

exports[`AtAccordion should render prop -- note 1`] = `
<view
  class="at-accordion"
>
  <view
    class="at-accordion__header"
  >
    
    <view
      class="at-accordion__info"
    >
      <view
        class="at-accordion__info__title"
      />
      <view
        class="at-accordion__info__note"
      >
        note
      </view>
    </view>
    <view
      class="at-accordion__arrow"
    >
      <text
        class="at-icon at-icon-chevron-down"
      />
    </view>
  </view>
  <view
    class="at-accordion__content at-accordion__content--inactive"
  >
    <view
      class="at-accordion__body"
      id="content"
    >
      按钮
    </view>
  </view>
</view>
`;

exports[`AtAccordion should render prop -- open=false 1`] = `
<view
  class="at-accordion"
>
  <view
    class="at-accordion__header"
  >
    
    <view
      class="at-accordion__info"
    >
      <view
        class="at-accordion__info__title"
      />
      <view
        class="at-accordion__info__note"
      />
    </view>
    <view
      class="at-accordion__arrow"
    >
      <text
        class="at-icon at-icon-chevron-down"
      />
    </view>
  </view>
  <view
    class="at-accordion__content at-accordion__content--inactive"
  >
    <view
      class="at-accordion__body"
      id="content"
    >
      按钮
    </view>
  </view>
</view>
`;

exports[`AtAccordion should render prop -- open=true 1`] = `
<view
  class="at-accordion"
>
  <view
    class="at-accordion__header"
  >
    
    <view
      class="at-accordion__info"
    >
      <view
        class="at-accordion__info__title"
      />
      <view
        class="at-accordion__info__note"
      />
    </view>
    <view
      class="at-accordion__arrow at-accordion__arrow--folded"
    >
      <text
        class="at-icon at-icon-chevron-down"
      />
    </view>
  </view>
  <view
    class="at-accordion__content"
  >
    <view
      class="at-accordion__body"
      id="content"
    >
      按钮
    </view>
  </view>
</view>
`;

exports[`AtAccordion should render prop -- title 1`] = `
<view
  class="at-accordion"
>
  <view
    class="at-accordion__header"
  >
    
    <view
      class="at-accordion__info"
    >
      <view
        class="at-accordion__info__title"
      >
        title
      </view>
      <view
        class="at-accordion__info__note"
      />
    </view>
    <view
      class="at-accordion__arrow"
    >
      <text
        class="at-icon at-icon-chevron-down"
      />
    </view>
  </view>
  <view
    class="at-accordion__content at-accordion__content--inactive"
  >
    <view
      class="at-accordion__body"
      id="content"
    >
      按钮
    </view>
  </view>
</view>
`;
